File: 프로그래머스/3/152995. 인사고과/인사고과.kt
class Solution {
    fun solution(orogin_scores: Array<IntArray>): Int {
        val wanhoScore = orogin_scores.first().sum()

        val scores = orogin_scores.sortedWith(compareBy( { -it.first() }, { it[1] }))
        
        var maxScore2 = 0
        var answer = 1
        for ((score1, score2) in scores) {
            if (orogin_scores[0][0] < score1 && orogin_scores[0][1] < score2) {
                return -1
            }
            if (score2 < maxScore2) continue
            if (wanhoScore < score1 + score2) {
                answer += 1
            }
            maxScore2 = score2
        }
        
        return answer
    }
}

File: 프로그래머스/3/152995. 인사고과/인사고과.py
LIMIT = 100_000

def solution(scores):
    max_score2 = [0 for _ in range(LIMIT+1)]
    for score1, score2 in scores:
        if score2 > max_score2[score1]:
            max_score2[score1] = score2
            
    # 얘가 기준: 해당 idx 일 때 value 보다 크거나 같으면 인센티브를 받는다.
    min_score2 = [0 for _ in range(LIMIT+1)]
    behind_max = 0
    for i in range(LIMIT, -1, -1):
        min_score2[i] = behind_max
        behind_max = max(behind_max, max_score2[i])
        
    live_scores = []
    for score1, score2 in scores[1:]:
        if score2 >= min_score2[score1]:
            live_scores.append(score1 + score2)
            
    if scores[0][1] < min_score2[scores[0][0]]:
        return -1
    
    stand = sum(scores[0])
    answer = 0
    for score in live_scores:
        if score > stand:
            answer += 1
    return answer + 1